File: backend/core/ids.py
import re
import secrets
import string
from datetime import date

from .models import Lecturer, Student

DEGREE_PREFIXES = {
    "PROFESSIONAL DIPLOMA": "PD",
    "DIPLOMA": "PD",
    "HND": "HND",
}

BACHELOR_MARKERS = ("BSC", "BBA", "BENG", "BTECH", "BA", "B.A.", "BACHELOR")

PROGRAM_AREA_CODES = {
    "INFORMATION TECHNOLOGY": "IT",
    "INFORMATION SYSTEMS": "IS",
    "COMPUTER SCIENCE": "CS",
    "NETWORK ENGINEERING": "NET",
    "NETWORKING": "NET",
    "ACCOUNTING": "ACC",
    "ENTREPRENEURSHIP": "ENT",
    "ELECTRICAL": "EEE",
    "ELECTRONIC": "EEE",
    "BUSINESS": "BUS",
}

STAFF_PREFIX = "LC"


def generate_temp_password(length=8):
    chars = string.ascii_letters + string.digits
    while True:
        password = "".join(secrets.choice(chars) for _ in range(length))
        if any(c.isalpha() for c in password) and any(c.isdigit() for c in password):
            return password


def _area_code(program=None, department=None):
    source = ""
    if program is not None and getattr(program, "name", None):
        source = program.name
    elif department is not None and getattr(department, "name", None):
        source = department.name
    upper = source.upper()
    for key, code in PROGRAM_AREA_CODES.items():
        if key in upper:
            return code
    words = [w for w in re.split(r"[^A-Za-z]+", upper) if w]
    if not words:
        return "GEN"
    return "".join(w[0] for w in words)[:3]


def degree_prefix(program=None):
    name = (program.name if program is not None else "").upper()
    for marker, prefix in DEGREE_PREFIXES.items():
        if marker in name:
            return prefix
    if name and any(m in name for m in BACHELOR_MARKERS):
        return "BC"
    return "BC"


def admission_year(created_at=None):
    return str((created_at or date.today()).year)[-2:]


def generate_university_email(identifier):
    slug = (identifier or "").replace("/", "").upper()
    if not slug:
        slug = f"CDAR{secrets.randbelow(1000000):06d}"
    return f"{slug}@edu.com"


def _next_sequence(model, field, base):
    existing = model.objects.filter(**{f"{field}__startswith": base + "/"}).values_list(field, flat=True)
    numbers = []
    for value in existing:
        if not value:
            continue
        tail = value.rsplit("/", 1)[-1]
        if tail.isdigit():
            numbers.append(int(tail))
    return (max(numbers) + 1) if numbers else 1


def generate_student_id(department, program, created_at=None):
    prefix = degree_prefix(program)
    area = _area_code(program, department)
    year = admission_year(created_at)
    base = f"{prefix}/{area}/{year}"
    seq = _next_sequence(Student, "student_id", base)
    return f"{base}/{seq:02d}"


def generate_staff_id(department):
    area = _area_code(department=department)
    base = f"{STAFF_PREFIX}/{area}"
    seq = _next_sequence(Lecturer, "staff_id", base)
    return f"{base}/{seq:02d}"

File: server/src/ids.js
const DEGREE_PREFIXES = {
  "PROFESSIONAL DIPLOMA": "PD",
  DIPLOMA: "PD",
  HND: "HND",
};

const BACHELOR_MARKERS = ["BSC", "BBA", "BENG", "BTECH", "BA", "B.A.", "BACHELOR"];

const PROGRAM_AREA_CODES = {
  "INFORMATION TECHNOLOGY": "IT",
  "INFORMATION SYSTEMS": "IS",
  "COMPUTER SCIENCE": "CS",
  "NETWORK ENGINEERING": "NET",
  NETWORKING: "NET",
  ACCOUNTING: "ACC",
  ENTREPRENEURSHIP: "ENT",
  ELECTRICAL: "EEE",
  ELECTRONIC: "EEE",
  BUSINESS: "BUS",
};

const STAFF_PREFIX = "LC";

export function generateTempPassword(length = 8) {
  const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  let password;
  do {
    password = "";
    for (let i = 0; i < length; i += 1) {
      password += chars[Math.floor(Math.random() * chars.length)];
    }
  } while (!/[a-zA-Z]/.test(password) || !/[0-9]/.test(password));
  return password;
}

export function areaCode(program = null, department = null) {
  let source = "";
  if (program && program.name) source = program.name;
  else if (department && department.name) source = department.name;
  const upper = source.toUpperCase();
  for (const key of Object.keys(PROGRAM_AREA_CODES)) {
    if (upper.includes(key)) return PROGRAM_AREA_CODES[key];
  }
  const words = upper.split(/[^A-Za-z]+/).filter(Boolean);
  if (!words.length) return "GEN";
  return words
    .map((w) => w[0])
    .join("")
    .slice(0, 3);
}

export function degreePrefix(program = null) {
  const name = (program ? program.name : "").toUpperCase();
  for (const marker of Object.keys(DEGREE_PREFIXES)) {
    if (name.includes(marker)) return DEGREE_PREFIXES[marker];
  }
  if (name && BACHELOR_MARKERS.some((m) => name.includes(m))) return "BC";
  return "BC";
}

export function admissionYear(createdAt = null) {
  const year = (createdAt ? new Date(createdAt).getFullYear() : new Date().getFullYear()).toString();
  return year.slice(-2);
}

export function generateUniversityEmail(identifier) {
  let slug = (identifier || "").replace(/\//g, "").toUpperCase();
  if (!slug) slug = `CDAR${String(Math.floor(Math.random() * 1000000)).padStart(6, "0")}`;
  return `${slug}@edu.com`;
}

function nextSequence(rows, field, base) {
  let max = 0;
  for (const row of rows) {
    const value = row[field];
    if (!value || !value.startsWith(`${base}/`)) continue;
    const tail = value.split("/").pop();
    if (/^\d+$/.test(tail)) max = Math.max(max, Number(tail));
  }
  return max + 1;
}

export function generateStudentId(students, department, program, createdAt = null) {
  const prefix = degreePrefix(program);
  const area = areaCode(program, department);
  const year = admissionYear(createdAt);
  const base = `${prefix}/${area}/${year}`;
  const seq = nextSequence(students, "student_id", base);
  return `${base}/${String(seq).padStart(2, "0")}`;
}

export function generateStaffId(lecturers, department) {
  const area = areaCode(null, department);
  const base = `${STAFF_PREFIX}/${area}`;
  const seq = nextSequence(lecturers, "staff_id", base);
  return `${base}/${String(seq).padStart(2, "0")}`;
}